### PR TITLE
Add Dict.assoc

### PR DESCRIPTION
**Pull request by anthony-khong**
Hi @mobily, this is my first attempt regarding #7. Of course, I still don't know if you actually want this in this library, but here is a first attempt anyway.

I have a couple of concerns/questions here:
* I'm still fairly new with TS - is the type declaration okay? In particular, I'm not sure about the use of mapped type in the return, and whether we should `Omit` the original key. I imagine `D.assoc({a: 1}, 'a', 'abc')` should be of type `{a: string}`.
* I've never used Rescript before - is the current implementation okay? I tried doing something along the lines of `merge(dict, {str(key): value})` but it doesn't type check or I get a syntax error.

### CODE REVIEW

File: src/Dict/Dict.res [hunk 1]
```diff
@@ -23,6 +23,13 @@ export fromPairs = dict => Js.Dict.fromArray(dict)
 %comment("Merges two provided objects.")
 export merge = (fst, snd) => assign(makeEmpty(), fst, snd)
 
+%comment("Adds a property. Equivalent to merging with `{key: value}`")
+export set = (dict, key, value) => {
+  let pairs = toPairs(dict)
+  let appended = Array.append(pairs, (key, value))
+  fromPairs(appended)
+}
```
Comment (mobily) [inline src/Dict/Dict.res:31]: looks good, although, I'd use this solution (I believe there's no need to iterate):

```rescript
export set = (dict, key, value) => {
  let obj = assign(makeEmpty(), dict)
  Js.Dict.set(obj, key, value)
  obj
}
```

File: src/Dict/Dict.ts [hunk 2]
```diff
@@ -7,6 +7,11 @@ export declare const keys: <T extends Record<string, unknown>>(
   dict: T,
 ) => ReadonlyArray<keyof T>
 export declare const merge: <A, B>(fst: A, snd: B) => A & B
+export declare const set: <T, K extends string, V>(
+  dict: T,
+  key: K,
+  value: V,
+) => T & { [key in K]: V }
```
Comment (mobily) [inline src/Dict/Dict.ts:14]: looks good as well! I think we should extend `T` with `string | number` and add tests for both cases 

```ts
export declare const set: <T, K extends string | number, V>(
  dict: T,
  key: K,
  value: V,
) => T & Record<K, V>
```

**Review by mobily — COMMENTED** [review]
so sorry for the delayed response! 🙈

**Comment by anthony-khong** [comment]
Hi @mobily! Thank you so much for looking into this and teaching a lesson in ReScript 🙈 I'll comment directly on the threads!

File: src/Dict/Dict.res (hunk 1 above)
Comment (anthony-khong) [reply src/Dict/Dict.res:31]: Yes, that's really nice. I believe you mean `merge` instead of  `assign`?

Comment (mobily) [reply src/Dict/Dict.res:31]: good point, both are ok 💯

**Pull request merged**